nx/lib/nx/lin_alg.ex

```diff
@@ -1373,50 +1403,44 @@ defmodule Nx.LinAlg do
   # We need a special-case for 0 since the code below
   # is optimized to not compute an initial eye.
   def matrix_power(tensor, 0) do
-    case Nx.shape(tensor) do
-      {n, n} ->
-        :ok
+    shape = Nx.shape(tensor)
 
-      shape ->
-        raise ArgumentError,
-              "matrix_power/2 expects a square tensor, got tensor with shape: #{inspect(shape)}"
+    with :ok <- Nx.Shape.matrix_power(shape) do
+      Nx.eye(tensor)
     end
-
-    Nx.eye(tensor)
   end
 
   def matrix_power(tensor, power) when is_integer(power) do
-    case Nx.shape(tensor) do
-      {n, n} ->
-        :ok
+    shape = Nx.shape(tensor)
 
-      shape ->
-        raise ArgumentError,
-              "matrix_power/2 expects a square tensor, got tensor with shape: #{inspect(shape)}"
+    with :ok <- Nx.Shape.matrix_power(shape) do
```

Comment:
In this case, you could write `:ok = Nx.Shape.matrix_power(shape)` without `with`, but it is a small change I can fix locally. :)